proposals/2024-07-15_qml_mapping_requirements.md

```diff
@@ -0,0 +1,84 @@
+## QML mapping requirements
+
+* **Owners:**
+  * Daniel Schürmann
+
+* **Implementation Status:** `Not implemented`
+
+* **Related Issues and PRs:**
+  * Create a reactive programming API for controllers [#13440](https://github.com/mixxxdj/mixxx/issues/13440) 
+  * QML-based components API for controllers [#13459](https://github.com/mixxxdj/mixxx/pull/13459)
+  * Respect the Midi timestamp when scratching [#6951](https://github.com/mixxxdj/mixxx/issues/6951)
+  * make brake, soft start, and spinback part of the effects system [#8867](https://github.com/mixxxdj/mixxx/issues/8867)
+
+* **Other docs or links:**
+  * `<Links…>`
+
+* ** This document shall collect requirements and expectation a new mapping system based on QMl shall fullfill and shall not fulfill. 
+
+## Why
+
+It’s important to clearly explain the reasons behind certain design decisions in order to have a consensus
+between team members, as well as external stakeholders. 
+
+It turned out that we have conflicting or unclear requirements. We need to decide the way forward to not end up with a blocked PR or double work.
+We need to decide which of them are mandatory optional.
+
+### Pitfalls of the current solution
+
+TODO
+
+## Goals
+
+Here is a fist draft of requirements
+
+1. User Requirements
+
+1.1 Good performance, responsive scratching (Real-Time)
+1.2 Easy reassign buttons and sliders
+1.3 Easy map a new controller from the scratch
+
+2. Mapping Developer
+
+2.1 Everything at one place, one file, folder, pack
+2.2     Easy to share
+2.3 Reuse common programming skills
+2.4 High abstraction level
+2.5 Complete API documentation
+2.6 Good IDE support .. code completion.
+2.7 On the fly edits, no Mixxx restart.
+2.8 Stable: Compatible to future Mixxx versions
+
```

Comment:
>
> 1. User Requirements

> 1.1 Good performance, responsive scratching (Real-Time)

I believe this can be achieved, however scratching is IMO orthogonal to QML. The missing piece here is the ability to take message timestamps into account, which has nothing to do with QML perse and is mostly an API extension in general. 

> 1.2 Easy reassign buttons and sliders

I think this depends on the definition of easy. If the mapping is written in a Components-oriented fashion, it won't be just simple point-and-click, but we can still retain the C++ dispatcher code that is point-and-click friendly.


> 1.3 Easy map a new controller from the scratch

This also hinges on the definition of easy. Knowing QML and MIDI fairly well, I'd say its easy. For someone that has never opened a text editor with code, not so much.

> 2.1 Everything at one place, one file, folder, pack
> 2.2     Easy to share

That should be fairly achievable for QML, however it does require a little bit of automated work to make the mapping "export-ready". So the current plain copying files wouldn't work and instead we'd need an export button that prepares the mapping accordingly. That prepwork would primarily consist of ensuring the built-in shared modules are bundled correctly.

> 2.3 Reuse common programming skills
> 2.4 High abstraction level
> 2.5 Complete API documentation
> 2.6 Good IDE support .. code completion.
> 2.7 On the fly edits, no Mixxx restart.
> 2.8 Stable: Compatible to future Mixxx versions

Should all be possible, especially when we can take advantage of QMLs built-in API versioning mechanics. IDE support hinges a little bit on the maturity of the qml-language-server.